src/hotspot/cpu/x86/c1_LIRAssembler_x86.cpp

```diff
@@ -3629,13 +3629,27 @@ void LIR_Assembler::emit_profile_type(LIR_OpProfileType* op) {
 
   __ verify_oop(obj);
 
-  if (tmp != obj) {
-    __ mov(tmp, obj);
+#ifdef ASSERT
+  if (obj == tmp) {
+    LP64_ONLY(assert_different_registers(obj, rscratch1, mdo_addr.base(), mdo_addr.index());)
```

Comment:
I mind `#ifdef _LP64` more readable than `LP64_ONLY`/`NOT_LP64` here.

Reply:
Moreover, any particular problems with unconditionally passing `rscratch1`? `rscratch1` is set to`noreg` on x86_32 and I believe `assert_different_registers` should handle it well if all other arguments are guaranteed to be valid registers.

Reply:
mdo_addr.index() can be noreg.
I used LP64_ONLY/NOT_LP64 because it was more compact, but I'm OK with #ifdef too.